### PR TITLE
Cleanlab wrapper

### LINKED ISSUE #223: cleanlab wrapper

### PR DESCRIPTION
**Pull request by agsedova**
close #223

### CODE REVIEW

**Comment by agsedova** [comment]
WARNING! do not merge before https://github.com/knodle/knodle/pull/238

File: knodle/transformation/torch_input.py [hunk 1]
```diff
@@ -26,3 +26,12 @@ def input_info_labels_to_tensordataset(
     input_ids_label_dataset = TensorDataset(*model_tensors, torch.from_numpy(input_info), torch.from_numpy(labels))
 
     return input_ids_label_dataset
+
+
+def dataset_to_numpy_input(model_input_x: TensorDataset) -> np.ndarray:
+    if len(model_input_x.tensors) == 1:
+        return model_input_x.tensors[0].numpy()
+    else:
+        raise ValueError(f"Selected denoising method accepts only two-dimensional encoded input features (features x "
```
Comment (marina-sp) [inline knodle/transformation/torch_input.py:35]: The error message does not correspond to the "if" condition. The message reports on the dimensionality of the input, however the "if" check is about how many different matrices/tensors are available.

**Review by marina-sp — CHANGES_REQUESTED** [review]

File: knodle/trainer/cleanlab/config.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+
+@AutoConfig.register("cleanlab")
+class CleanLabConfig(MajorityConfig):
+    def __init__(
+            self,
+            cv_n_folds=5,
+            prune_method='prune_by_noise_rate',
+            converge_latent_estimates=False,
+            pulearning=None,
```
[the remark this answers is not in the document]
Comment (marina-sp) [reply knodle/trainer/cleanlab/config.py:13]: Marked too many rows, this is related to "pulearning" parameter

File: examples/trainer/cleanlab/cleanlab_training_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    df_train, df_dev, df_test, train_rule_matches_z, _, mapping_rules_labels_t = read_train_dev_test(path_to_data)
+
+    train_input_x, test_input_x, _ = get_tfidf_features(df_train["sample"], test_data=df_test["sample"])
+
+    train_features_dataset = TensorDataset(Tensor(train_input_x.toarray()))
+    test_features_dataset = TensorDataset(Tensor(test_input_x.toarray()))
+
+    test_labels = list(df_test.iloc[:, -1])
+    test_labels_dataset = TensorDataset(LongTensor(list(df_test.iloc[:, 1])))
```
[the remark this answers is not in the document]
Comment (agsedova) [reply examples/trainer/cleanlab/cleanlab_training_tutorial.py:35]: but e.g. in tac dataset the column "label" contains labels as words, while the label indices (which should be used here) are in the last column called "enc_label". 

Another possible solution would be to rename the tac columns so that the column "label" would contain label ids, while the labels-words would be in another column (somehow "text labels"). Do you think it would be a better option?

File: knodle/transformation/torch_input.py (hunk 1 above)
Comment (agsedova) [reply knodle/transformation/torch_input.py:35]: done

File: knodle/trainer/trainer.py [hunk 4]
```diff
@@ -242,10 +242,14 @@ def test(
             self, features_dataset: TensorDataset, labels: TensorDataset, loss_calculation: bool = False
     ) -> Tuple[Dict, Union[float, None]]:
 
-        feature_label_dataset = input_labels_to_tensordataset(features_dataset, labels.tensors[0].cpu().numpy())
-        feature_label_dataloader = self._make_dataloader(feature_label_dataset, shuffle=False)
+        gold_labels = labels.tensors[0].cpu().numpy()
 
-        predictions, gold_labels, dev_loss = self._prediction_loop(feature_label_dataloader, loss_calculation)
+        if hasattr(self.model, "predict"):       # when the pytorch model is wrapped as a sklearn model (e.g. cleanlab)
```
[the remark this answers is not in the document]
Comment (agsedova) [reply knodle/trainer/trainer.py:247]: agree with `if isinstance(self.model, skorch.NeuralNetClassifier)`, but would rather have this loop here since the rest of the test function works for skorch classifier.

File: knodle/trainer/cleanlab/config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (agsedova) [reply knodle/trainer/cleanlab/config.py:13]: this is the cleanlab parameter, and since it is a simple wrapper, I would not rename it here.
I would add a more detailed explanation here.

File: examples/trainer/cleanlab/cleanlab_training_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import sys
+from itertools import product
+
+from torch import Tensor, LongTensor
+from torch.nn import CrossEntropyLoss
+from torch.optim import Adam
+from torch.utils.data import TensorDataset
+
+from examples.trainer.dscrossweigh.dscrossweigh_training_tutorial import get_tfidf_features
+from examples.utils import read_train_dev_test
```
[the remark this answers is not in the document]
Comment (agsedova) [reply examples/trainer/cleanlab/cleanlab_training_tutorial.py:12]: the function is relevant only to the tutorials, that's why it is in examples.utils

[the remark this answers is not in the document]
Comment (agsedova) [reply examples/trainer/cleanlab/cleanlab_training_tutorial.py:12]: ... and line 11 is changed now to `from examples.trainer.preprocessing import get_tfidf_features`:)
